chapter/1/rpc-reviewed-dont-keep.md

```diff
@@ -0,0 +1,619 @@
+Invocation) model to facilitate outsourcing computation. Cuckoo {% cite cuckoo
+--file rpc %} uses this Ibis communication middleware to offload computation.
+
+The Microsoft's MAUI Project {% cite maui --file rpc %} uses RPC communication
+and allows partitioning of .NET applications and "fine-grained code offload to
+maximize energy savings with minimal burden on the programmer". MAUI decides the
+methods to offload to the external MAUI server at runtime.
+
+#### Async RPC, Futures and Promises
+
+Remote Procedure Calls can be asynchronous. Not only that but these async RPCs
+play in integral role in the *futures* and *promises*. *Future* and *promises*
+are programming constructs that where a *future* is seen as variable/data/return
+type/error while a *promise* is seen as a *future* that doesn't have a value,
+yet. We follow Finagle's {% cite finagle --file rpc %} definition of *futures*
+and *promises*, where the *promise* of a *future*(an empty *future*) is
+considered as a *request* while the async fulfillment of this *promise* by a
+*future* is seen as the *response*. This construct is primarily used for
+concurrent programming.
+
+Perhaps the most renowned systems using this type of RPC model are Twitter's
+Finagle{% cite finagle --file rpc %} and Cap'n Proto{% cite capnproto --file rpc %}.
+
+#### RPC in Microservices Ecosystem:
+
+RPC implementations have moved from a one-server model to multiple servers and
+on to dynamically-created, load-balanced microservices. RPC started as a
+separate implementations of REST, Streaming RPC, MAUI, gRPC, Cap'n Proto, and
+has now made it possible for integration of all these implementations as a
+single abstraction as a user *endpoint* service. The endpoints are the building
+blocks of *microservices*. These *microservices* interact with each other and
+applications and combine to give the feel of one large monolithic service.
+
+The use of RPC has allowed us to create new microservices on-the-fly. The
+microservices can not only created and bootstrapped at runtime but also have
+inherent features like load-balancing and failure-recovery. This bootstrapping
+might occur on the same machine, adding to a Docker container {% cite docker
+--file rpc %}, or across a network (using any combination of DNS, NATs or other
+mechanisms).
+
+RPC can be defined as the "glue" that holds all the microservices together{%
+cite microservices1rpc --file rpc %}. This means that RPC is one of the primary
+communication mechanism between different microservices running on different
+systems. A microservice requests another microservice to perform an
+operation/query. The other microservice, upon receiving such request, performs
+an operation and returns a response. This operation could vary from a simple
+computation to invoking another microservice creating a series of RPC events to
+creating new microservices on the fly to dynamically load balance the
+microservices system.
```

Comment:
I would make points somewhere in this section about it being possible for microservices to each be written in different languages by totally different teams of developers but to be able to use language-agnostic RPC frameworks like gRPC or Thrift to that are performant (asynchronous + optimized serialization). This point isn't directly made anywhere.